README
This module wraps the MPC library functions.

This version of Math::MPC needs mpc-0.8 or later.

The MPC library is a C library for multiple-precision complex number
computations with exact rounding. It is based on the MPFR C library 
which, in turn, is based on the GMP C library.

MPC, like MPFR and GMP, is free. It is distributed under the
GNU Lesser General Public License (GNU Lesser GPL).

Because this module wraps the MPC functions, it requires that the
MPC library (version 0.8.0 or later), the GMP library (version 4.3.2
or later) and the MPFR library (version 2.4.2 or later) have been
installed. 

For the GMP library see:
http://gmplib.org

For the MPFR library see:
http://www.mpfr.org
Do not use the (outdated) MPFR library source that ships with the
GMP source.

For the MPC library see:
http://www.multiprecision.org/mpc/

You'll also need to have Math::MPFR (version 2.01 or later) installed.

To build this module you need perl 5.6.0 or later. I'm not sure of all
the requirements to get it to build with earlier versions of perl, but
it's definite that the XS code relating to operator overloading will
not compile with perl versions prior to 5.6.

Build in the usual way:

 perl Makefile.PL
 make
 make test
 make install

When building this module, the MPC, GMP and MPFR libraries will need to be
accessible. If those files are in a location where your compiler does 
not find them by default, then instead of running 'perl Makefile.PL', 
you'll need to run:

perl Makefile.pl INC="-I/path/to/gmp_includes -I/path/to/mpfr_includes -I/path/to/mpc/include"
 LIBS="-L/path/to/mpc_lib -lmpc -L/path/to/mpfr_lib -lmpfr -L/path/to/gmp_lib -lgmp"

==============
64-bit support
==============

If your perl's Config reports that 'ivsize' is greater than 
or equal to 8,then Math::MPC will, by default, be built with access
to the mpfr_*_uj and mpfr_*_sj functions. Else, access to those 
functions is, by default, denied. You can override the default by 
opening up the Makefile.PL and uncommenting the appropriate line (just
a few lines down from the top of the file ... the comments in that
file should make it clear).

Similarly, if your perl's Config reports that 'nvsize' is greater than
8 then Math::MPC will, by default, be built with access
to the mpfr_*_ld functions. Else, access to those functions is, 
by default, denied. You can override the default by opening up the 
Makefile.PL and uncommenting the appropriate line (just a few lines down
from the top of the file ... the comments in that file should make it
clear).

I believe it to be both unnecessary and inadvisable to override the 
default - but if you do find that you need to override the default, 
please let me know. That you have found it necessary to override the
default is something I probably ought to know about.

=========================
_Complex C types support
=========================

The mpc library (beginning with version 0.9) provides support for assigning
and returning the 'double _Complex' and 'long double _Complex' C data types.
Therefore Math::MPC (beginning with version 0.90) provides optional support for
assigning/returning Math::Complex_C (double _Complex) objects and Math::Complex_C::Long
(long double _Complex) objects, using the Rmpc_set_dc, Rmpc_set_ldc, Rmpc_get_dc and
Rmpc_get_ldc functions.

If Math::Complex_C can be loaded then complex.h will be included in the compilation
and, iff _MPC_H_HAVE_COMPLEX is thereby defined (in mpc.h), then Math::MPC will be built
with support for the above mentioned set/get functions.

If there's a need to override this default behaviour, just open the Makefile.PL and
comment in the line of code that forces '$do_complex_h' to be assigned the value you
want ... either -1 (to force exclusion of support), or 1 (to force support). Then
run 'make realclean' and start over again.

Any problems, let me know.

I am contactable by email at sisyphus at(@) cpan dot (.) org.




CHANGES
1.01
 - Alter the way that strEQ gets hold of HvNAME(SvSTASH(SvRV(sv))) - so that -Wall
   doesn't emit "null argument" warnings. (Thanks Daniel Kahn Gillmor.)
 - Alter test 5 of t/complex_c.t to accommodate version 0.07 (and later) of
   Math::Complex_C.

1.00
 - Add Rmpc_log10, Rmpc_mul2si and Rmpc-div_2si (new in 1.0.0)
 - Accommodate namechange of mpc_mul_2exp and mpc_div_2exp to mpc_mul_2ui
   and mpc_div_2ui (respectively) in mpc-1.0.0.
 - Rewrite the detecting of availability of double _Complex and long double _Complex
   types by looking for _Complex_I instead of _MPC_H_HAVE_COMPLEX (in line with
   the way mpc.h currently does it).

0.93
 - Remove 'not' from overloads. Apparently achieved nothing anyway, and 5.16
   warns that it's invalid.

0.92
 - The default rounding mode and default precision globals are now thread-safe.
   Consequently at least 5.008008 is required.

0.91
 - overload_equiv and overload_not_equiv weren't correctly handling the
   case where the NV is a NaN. (Add tests in mpc_overload.t.)

0.90
 - Add Rmpc_sin_cos, Rmpc_set_dc, Rmpc_set_ldc, Rmpc_get_dc, Rmpc_get_ldc
   (new in mpc-0.9).
 - Add documentation of gmp_v and mpfr_v (previously missed).
 - In the documentation, elaborate on the way that Rmpc_get-version() and 
   MPC_VERSION & friends determine their values. (Thanks Vincent Lefevre.) 
 - In MPC.xs change mpfr_v() to return mpfr_get_version() instead of
   MPC_VERSION_STRING.
 - In t/aab_basic.t print out the mpfr and mpc library versions from 
   the libraries, instead of from the headers (as had been the case)
 - Some of the croak() messages in MPC.xs finished with "\n" - which
   has now been removed from those messages.
 - Releases of the MPC library (after 0.8.2) will require gmp-4.3.2
   and mpfr-2.4.2. README and Makefile.PL have now been changed.

0.83
 - Add some warn() statements to t/bool.t, to try and see why test 2
   fails - as reported by one of the cpan-testers. (That done, I
   received an email from Jean-Louis Morel, with a patch that fixes
   the issue ... I'll retain t/bool.t in its modified state.)
 - Patch overload_not, overload_not_equiv and overload_equiv (in
   MPC.xs) so that NaNs do not get passed on to the MPC comparison
   functions (as this is something forbidden by the documentation).
   Thanks to Jean-Louis Morel.

0.82
 - Implement '#include <inttypes.h>'. (Should have been done for 0.81.)
   See https://rt.cpan.org/Ticket/Display.html?id=58564

0.81
 - Change the overload_string() and _get_str() functions to match the
   formatting provided by mpc_out_str().
 - Fix the signed zero problem with overloaded mul and div operations,
   and add t/signed_zero.t as a check. Thanks Paul Zimmerman and 
   Philippe Theveny.
 - Add Rmpc_mul_sj, Rmpc_mul_ld, Rmpc_mul_d, Rmpc_div_sj, Rmpc_sj_div,
   Rmpc_div_ld, Rmpc_ld_div, Rmpc_div_d and Rmpc_d_div. These simply
   wrap the the XSubs created to fix the signed zero anomaly (above). 

0.80
 - Requires version 0.8 of the mpc library.
 - Add Rmpc_asin, Rmpc_acos, Rmpc_atan, Rmpc_asinh, Rmpc_acosh and
   Rmpc_atanh (new in mpc-0.8).
 - Add Rmpc_pow_d, Rmpc_pow_ld, Rmpc_pow_si, Rmpc_pow_ui, Rmpc_pow_z
   and Rmpc_pow_fr (new in mpc-0.8).
 - Add overloaded atan2 function.

0.70
 - Requires version 0.7 of the mpc library. (Change to MPC_SET_X_Y.)
 - Remove Rmpc_random and Rmpc_random2 (as the functions that they
   wrap have been removed in mpc-0.7. (And remove t/mpc_random.t.)
 - Add Rmpc_pow, Rmpc_set_nan and Rmpc_swap (new in mpc-0.7).
 - Overload the ** and **= operators.
 - Improve the efficiency of some of the overload routines by
   eliminating the need to use a temporary mpfr object (using the
   already existing *mpc_t_obj instead).
 - Remove the rounding arg from RMPC_RE() and RMPC_IM() as it does
   nothing, anyway.

0.60
 - Requires version 0.6 of the mpc library.
 - Now a complete range of Rmpc_set_*_* functions (where '*' can be any
   one of 'ui', 'si', 'd', 'uj', 'sj', 'ld', 'f', 'q', 'z' and 'fr').
 - Add Rmpc_set_z, Rmpc_set_q, Rmpc_set_f and Rmpc_set_fr.
 - Add Rmpc_set_z_z, Rmpc_set_q_q and Rmpc_set_f_f.
 - In keeping with changes to the mpc library, all of the Rmpc_init*
   functions have been removed except for Rmpc_init2 and Rmpc_init3
   (and their "_nobless" forms).
 - mpc-0.60 has removed mpc_get_default_prec and mpc_set_default_prec.
   Rmpc_get_default_prec and Rmpc_set_default_prec remain (but the
   former has changed to return 0 if default real and imaginary
   precisions are different. With earlier versions, real and imaginary
   default precisions were always the same, but this is no longer the
   case). Rmpc_get_default_prec2 and Rmpc_set_default_prec2 have now
   been added. The former returns an array of 2 values(real,
   imaginary) and the latter requires 2 arguments (real, imaginary).
 - The overload_copy sub now preserves the precision of the copied object
   (instead of assigning default precision as was previously the case).
 - Rmpc_get_str now wraps mpc_get_str (available in 0.6). The (pure perl)
   sub that was formerly "Rmpc_get_str" has been renamed "_get_str".
   The "overload_string" sub is unchanged.
 - Add Rmpc_set_uj, Rmpc_set_sj and Rmpc_set_ld (available in mpc-0.6).
 - Add Rmpc_strtoc and Rmpc_set_str (available in 0.6), and add set_str.t
   test script.

0.52
 - Requires version 0.5.2 of the mpc library.
 - Add MPC_VERSION, MPC_VERSION_NUM and Rmpc_get_version.

0.51
 - Requires version 0.5.1 of the mpc library.
 - Add MPC_VERSION_STRING, MPC_VERSION_MAJOR, MPC_VERSION_MINOR
   and MPC_VERSION_PATCHLEVEL.
 - Replace pre-existing Rmpc_set_fr_fr with a function that wraps
   the mpc_set_fr_fr function (which became available in 0.5.1)
 - Add Rmpc_real, Rmpc_imag, Rmpc_arg, Rmpc_proj (new in 0.5.1)

0.50
 - Requires version 0.50 of the mpc library.
 - Add Rmpc_log, Rmpc_cos, Rmpc_tan, Rmpc_sinh, Rmpc_cosh, and

Makefile.PL
use ExtUtils::MakeMaker;
use Config;
require 5.008008;

$do_complex_h = 0; # Let the Makefile.PL decide whether to support the double _Complex and
                   # long double _Complex data types.
#$do_complex_h = 1; # Force inclusion of the double _Complex and long double _Complex data types
#$do_complex_h = -1;# Force exclusion of the double _Complex and long double _Complex data types

unless ($do_complex_h) {
  eval {require Math::Complex_C;};
  $do_complex_h = $@ ? -1 : 1;
}

$use_64_bit_int = 0; # Let perl decide whether to include 64-bit 'long long' support
$use_long_double = 0;# Let perl decide whether to include 'long double' support

#$use_64_bit_int = -1; # Force exclusion of 64-bit 'long long' support
#$use_long_double = -1;# Force exclusion of 'long double' support

#$use_64_bit_int = 1; # Force inclusion of 64-bit 'long long' support
#$use_long_double = 1;# Force inclusion of 'long double' support

print "\nThis module requires the following C libraries:\n";
print " gmp-4.3.2 (or later)\n mpfr-2.4.2 (or later)\n mpc-0.8 (or later)\n\n";

my $defines = $] < 5.008 ? "-DOLDPERL" : "-DNEWPERL";

if($use_64_bit_int == -1) {}                                
elsif($use_64_bit_int == 1) {$defines .= " -DUSE_64_BIT_INT"}
else {
  unless($Config::Config{ivsize} < 8) {
    $defines .= " -DUSE_64_BIT_INT";
  }
}

if($use_long_double == -1) {}
elsif($use_long_double == 1) {$defines .= " -DUSE_LONG_DOUBLE"}
else {
  if($Config::Config{nvsize} > 8 ) {
    $defines .= " -DUSE_LONG_DOUBLE";
  }
}

if($do_complex_h == 1) { $defines .= " -D_DO_COMPLEX_H"}

$defines =~ /-DUSE_64_BIT_INT/ ? print "Building with 64-bit'long long' support\n" :
                                 print "Building without 64-bit 'long long' support\n";

print "If this is wrong, see the \"64-bit support\" section in the README\n\n";

$defines =~ /-DUSE_LONG_DOUBLE/ ? print "Building with 'long double' support\n" :
                                 print "Building without 'long double' support\n";

print "If this is wrong, see the \"64-bit support\" section in the README\n\n";

$defines =~ /-D_DO_COMPLEX_H/ ? print "Trying to build with 'double _Complex' & 'long double _Complex' support\n" :
                                 print "Building without 'double _Complex' & 'long double _Complex' support\n";

print "If this is wrong, see the \"_Complex C types support\" section in the README\n\n";

my %options = (       
  NAME => 'Math::MPC',
  AUTHOR => 'Sisyphus (sisyphus at (@) cpan dot (.) org)',
  ABSTRACT => 'Perl interface to the MPC (multi precision complex) library',
  PREREQ_PM => {'Math::MPFR' => '2.01'},
  DEFINE   => $defines,
  LIBS => [
    '-lmpc -lmpfr -lgmp'
  ],
  VERSION_FROM => 'MPC.pm',
  clean   => { FILES => 'out1.txt out2.txt out3.txt out4.txt out5.txt out6.txt out7.txt save_child_setting.txt' },
);

WriteMakefile(%options);

# Remove the Makefile dependency. Causes problems on a few systems.
sub MY::makefile { '' }

